### Q1: What is the start date and end date of the dataset?

In [24]:
import duckdb

# Connect to the database created by your dlt pipeline
conn = duckdb.connect('taxi_pipeline.duckdb')

# Use the correct column names for the Zoomcamp dataset
query = """
SELECT 
    MIN(trip_pickup_date_time) AS start_date, 
    MAX(trip_pickup_date_time) AS end_date 
FROM taxi_data.rides;
"""

print("\n--- Question 1 Results ---")
conn.sql(query).show()

conn.close()


--- Question 1 Results ---
┌──────────────────────────┬──────────────────────────┐
│        start_date        │         end_date         │
│ timestamp with time zone │ timestamp with time zone │
├──────────────────────────┼──────────────────────────┤
│ 2009-06-01 04:33:00-07   │ 2009-06-30 16:58:00-07   │
└──────────────────────────┴──────────────────────────┘



**Answer: `2009-06-01 to 2009-07-01`**

---

### Q2: What proportion of trips are paid with credit card?

In [ ]:
import duckdb

# Connect to the database
conn = duckdb.connect('taxi_pipeline.duckdb')

# SQL to calculate the proportion
# payment_type is a STRING in this dataset (e.g. 'Credit', 'Cash')
# use CAST as FLOAT to ensure we don't get integer division (0)
query = """
SELECT 
    CAST(SUM(CASE WHEN payment_type = 'Credit' THEN 1 ELSE 0 END) AS FLOAT) / COUNT(*) AS credit_card_proportion
FROM taxi_data.rides;
"""

print("\n--- Homework Q2: Credit Card Proportion ---")
conn.sql(query).show()

conn.close()


--- Homework Q2: Credit Card Proportion ---
┌────────────────────────┐
│ credit_card_proportion │
│         float          │
├────────────────────────┤
│                 0.2666 │
└────────────────────────┘



**Answer: `26.66%`**

---

### Q3: What is the total amount of money generated in tips?

In [28]:
import duckdb

# Connect to the database
conn = duckdb.connect('taxi_pipeline.duckdb')

# SQL to calculate the total tips (column is tip_amt in this dataset)
query = """
SELECT 
    ROUND(SUM(tip_amt), 2) AS total_tips_generated
FROM taxi_data.rides;
"""

print("\n--- Homework Q3: Total Tips Generated ---")
conn.sql(query).show()

conn.close()


--- Homework Q3: Total Tips Generated ---
┌──────────────────────┐
│ total_tips_generated │
│        double        │
├──────────────────────┤
│              6063.41 │
└──────────────────────┘



**Answer: `$6,063.41`**